plugin/session_view.py

```diff
@@ -173,42 +174,44 @@ def shutdown_async(self) -> None:
         if listener:
             listener.on_session_shutdown_async(self.session)
 
-    def diagnostics_key(self, severity: int) -> str:
-        return "lsp{}d{}".format(self.session.config.name, severity)
+    def diagnostics_key(self, severity: int, multiline: bool) -> str:
+        return "lsp{}d{}{}".format(self.session.config.name, "m" if multiline else "s", severity)
 
     def diagnostics_tag_scope(self, tag: int) -> Optional[str]:
         for k, v in DiagnosticTag.__dict__.items():
             if v == tag:
                 return 'markup.{}.lsp'.format(k.lower())
         return None
 
-    def present_diagnostics_async(self, flags: int) -> None:
-        data_per_severity = self.session_buffer.data_per_severity
+    def present_diagnostics_async(self) -> None:
         for severity in reversed(range(1, len(DIAGNOSTIC_SEVERITY) + 1)):
-            key = self.diagnostics_key(severity)
-            key_tags = {tag: '{}_tags_{}'.format(key, tag) for tag in DIAGNOSTIC_TAG_VALUES}
-            for key_tag in key_tags.values():
-                self.view.erase_regions(key_tag)
-            data = data_per_severity.get(severity)
-            if data is None:
-                self.view.erase_regions(key)
-            elif ((severity <= userprefs().show_diagnostics_severity_level) and
-                    (data.icon or flags != (sublime.DRAW_NO_FILL | sublime.DRAW_NO_OUTLINE))):
-                non_tag_regions = data.regions
-                for tag, regions in data.regions_with_tag.items():
-                    tag_scope = self.diagnostics_tag_scope(tag)
-                    # Trick to only add tag regions if there is a corresponding color scheme scope defined.
-                    if tag_scope and 'background' in self.view.style_for_scope(tag_scope):
-                        self.view.add_regions(key_tags[tag], regions, tag_scope, flags=sublime.DRAW_NO_OUTLINE)
-                    else:
-                        non_tag_regions.extend(regions)
-                self.view.add_regions(key, non_tag_regions, data.scope, data.icon, flags | sublime.DRAW_EMPTY)
-            else:
-                self.view.erase_regions(key)
+            self._draw_diagnostics(severity, DIAGNOSTIC_SEVERITY[severity - 1][4], False)
+            self._draw_diagnostics(severity, DIAGNOSTIC_SEVERITY[severity - 1][5], True)
         listener = self.listener()
         if listener:
             listener.on_diagnostics_updated_async()
 
+    def _draw_diagnostics(self, severity: int, flags: int, multiline: bool) -> None:
+        key = self.diagnostics_key(severity, multiline)
+        key_tags = {tag: '{}_tags_{}'.format(key, tag) for tag in DIAGNOSTIC_TAG_VALUES}
+        for key_tag in key_tags.values():
+            self.view.erase_regions(key_tag)
+        data = self.session_buffer.data_per_severity.get((severity, multiline))
+        if data is None:
+            self.view.erase_regions(key)
```

Comment:
I guess it would be fine to remove this if now and move the data check to the next one.